server: TestSQLStatCollection failed

[(server).TestSQLStatCollection failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2765595&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
=== RUN   TestSQLStatCollection
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection018715986
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:191: expected stats count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000303404 squared_diffs:0 > plan_lat:<mean:0.005526980666666667 squared_diffs:0.00016733435933430466 > run_lat:<mean:0.000422245 squared_diffs:1.0503194408000003e-08 > service_lat:<mean:0.03439540133333333 squared_diffs:1.213841587404674e-06 > overhead_lat:<mean:0.028142771666666667 squared_diffs:0.00019566232893620869 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:1 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > >  found count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000303404 squared_diffs:0 > plan_lat:<mean:0.005526980666666667 squared_diffs:0.00016733435933430466 > run_lat:<mean:0.0004222450000000001 squared_diffs:1.0503194408000003e-08 > service_lat:<mean:0.03439540133333333 squared_diffs:1.213841587404674e-06 > overhead_lat:<mean:0.028142771666666667 squared_diffs:0.00019566232893620869 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:0 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > > 
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection018715986
--- FAIL: TestSQLStatCollection (0.35s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSQLStatCollection PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSQLStatCollection.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(server).TestSQLStatCollection failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2765649&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
=== RUN   TestSQLStatCollection
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection422962414
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:191: expected stats count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000453545 squared_diffs:0 > plan_lat:<mean:0.005543260333333334 squared_diffs:0.00017723469438711271 > run_lat:<mean:0.000507318 squared_diffs:2.3684021911999993e-08 > service_lat:<mean:0.035508948666666665 squared_diffs:1.0550685652086613e-06 > overhead_lat:<mean:0.029004825333333335 squared_diffs:0.00020657398408856264 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:1 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > >  found count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000453545 squared_diffs:0 > plan_lat:<mean:0.005543260333333334 squared_diffs:0.00017723469438711271 > run_lat:<mean:0.000507318 squared_diffs:2.3684021911999993e-08 > service_lat:<mean:0.035508948666666665 squared_diffs:1.0550685652086613e-06 > overhead_lat:<mean:0.029004825333333335 squared_diffs:0.00020657398408856264 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:0 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > > 
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection422962414
--- FAIL: TestSQLStatCollection (0.43s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSQLStatCollection PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSQLStatCollection.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(server).TestSQLStatCollection failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2766203&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
=== RUN   TestSQLStatCollection
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection581169659
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:191: expected stats count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000213862 squared_diffs:0 > plan_lat:<mean:0.004524727666666667 squared_diffs:0.00011927427151732267 > run_lat:<mean:0.00028239266666666666 squared_diffs:2.639576632666665e-09 > service_lat:<mean:0.026226537333333334 squared_diffs:3.043004559046653e-07 > overhead_lat:<mean:0.021205555 squared_diffs:0.000131166180919178 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:1 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > >  found count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000213862 squared_diffs:0 > plan_lat:<mean:0.004524727666666667 squared_diffs:0.00011927427151732267 > run_lat:<mean:0.00028239266666666666 squared_diffs:2.639576632666665e-09 > service_lat:<mean:0.026226537333333338 squared_diffs:3.043004559046653e-07 > overhead_lat:<mean:0.021205555 squared_diffs:0.000131166180919178 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:0 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > > 
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection581169659
--- FAIL: TestSQLStatCollection (0.29s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestSQLStatCollection PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSQLStatCollection.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(server).TestSQLStatCollection failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2765586&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
=== RUN   TestSQLStatCollection
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection890515758
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:191: expected stats count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000801711 squared_diffs:0 > plan_lat:<mean:0.028315591 squared_diffs:0.004642374564185504 > run_lat:<mean:0.004294756666666667 squared_diffs:1.576997622166665e-07 > service_lat:<mean:0.179937873 squared_diffs:7.279307060857388e-05 > overhead_lat:<mean:0.14652581433333334 squared_diffs:0.005699117926426251 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:1 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > >  found count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000801711 squared_diffs:0 > plan_lat:<mean:0.028315591 squared_diffs:0.004642374564185504 > run_lat:<mean:0.004294756666666667 squared_diffs:1.576997622166665e-07 > service_lat:<mean:0.179937873 squared_diffs:7.279307060857388e-05 > overhead_lat:<mean:0.14652581433333334 squared_diffs:0.005699117926426251 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:0 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > > 
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection890515758
--- FAIL: TestSQLStatCollection (1.37s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSQLStatCollection PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSQLStatCollection.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(server).TestSQLStatCollection failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2766204&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
=== RUN   TestSQLStatCollection
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection712828627
    test_log_scope.go:74: use -show-logs to present logs inline
    stats_test.go:191: expected stats count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.000472471 squared_diffs:0 > plan_lat:<mean:0.022169010000000003 squared_diffs:0.002858962876061703 > run_lat:<mean:0.003578392 squared_diffs:8.674384200019998e-07 > service_lat:<mean:0.1440602033333333 squared_diffs:3.910023575013292e-05 > overhead_lat:<mean:0.11784033033333331 squared_diffs:0.0035945553453463246 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:1 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > >  found count:3 first_attempt_count:3 max_retries:0 legacy_last_err:"" legacy_last_err_redacted:"" num_rows:<mean:1 squared_diffs:0 > parse_lat:<mean:0.0004724710000000001 squared_diffs:0 > plan_lat:<mean:0.022169010000000003 squared_diffs:0.002858962876061703 > run_lat:<mean:0.003578392 squared_diffs:8.674384200019998e-07 > service_lat:<mean:0.1440602033333333 squared_diffs:3.910023575013292e-05 > overhead_lat:<mean:0.11784033033333331 squared_diffs:0.0035945553453463246 > sensitive_info:<last_err:"<redacted>" most_recent_plan_description:<name:"" > most_recent_plan_timestamp:<seconds:-62135596800 > > bytes_read:<mean:0 squared_diffs:0 > rows_read:<mean:0 squared_diffs:0 > exec_stats:<count:0 network_bytes:<mean:0 squared_diffs:0 > max_mem_usage:<mean:10240 squared_diffs:0 > contention_time:<mean:0 squared_diffs:0 > network_messages:<mean:0 squared_diffs:0 > max_disk_usage:<mean:0 squared_diffs:0 > > 
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSQLStatCollection712828627
--- FAIL: TestSQLStatCollection (1.12s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestSQLStatCollection PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSQLStatCollection.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/roachpb/app_stats.go
+++ b/pkg/roachpb/app_stats.go
@@ -148,7 +148,7 @@ func (s *StatementStatistics) Add(other *StatementStatistics) {
 }
 
 // AlmostEqual compares two StatementStatistics and their contained NumericStats
-// objects within an window of size eps.
+// objects within an window of size eps, ExecStats are ignored.
 func (s *StatementStatistics) AlmostEqual(other *StatementStatistics, eps float64) bool {
 	return s.Count == other.Count &&
 		s.FirstAttemptCount == other.FirstAttemptCount &&
@@ -161,8 +161,10 @@ func (s *StatementStatistics) AlmostEqual(other *StatementStatistics, eps float6
 		s.OverheadLat.AlmostEqual(other.OverheadLat, eps) &&
 		s.SensitiveInfo.Equal(other.SensitiveInfo) &&
 		s.BytesRead.AlmostEqual(other.BytesRead, eps) &&
-		s.RowsRead.AlmostEqual(other.RowsRead, eps) &&
-		s.ExecStats.AlmostEqual(&other.ExecStats, eps)
+		s.RowsRead.AlmostEqual(other.RowsRead, eps)
+	// s.ExecStats are deliberately ignored since they are subject to sampling
+	// probability and are not fully deterministic (e.g. the number of network
+	// messages depends on the range cache state).
 }
 
 // Add combines other into this ExecStats.
@@ -181,15 +183,4 @@ func (s *ExecStats) Add(other ExecStats) {
 	s.MaxDiskUsage.Add(other.MaxDiskUsage, execStatCollectionCount, other.Count)
 
 	s.Count += s.Count
-}
-
-// AlmostEqual compares two ExecStats and their contained NumericStats objects
-// within an window of size eps.
-func (s *ExecStats) AlmostEqual(other *ExecStats, eps float64) bool {
-	return s.Count == other.Count &&
-		s.NetworkBytes.AlmostEqual(other.NetworkBytes, eps) &&
-		s.MaxMemUsage.AlmostEqual(other.MaxMemUsage, eps) &&
-		s.ContentionTime.AlmostEqual(other.ContentionTime, eps) &&
-		s.NetworkMessages.AlmostEqual(other.NetworkMessages, eps) &&
-		s.MaxDiskUsage.AlmostEqual(other.MaxDiskUsage, eps)
 }

